File: my_problems/day71-dsa-backtracking-I/hw_q2_All_Unique_Permutations.java
import java.util.ArrayList;
import java.util.Arrays;

/* All Unique Permutations */

/**
 * Recursive relation: T(N) = N * T(N-1) + 1
 * T(N) = N * [(N-1) * T(N-2) + 1] + 1
 * = N * (N-1) * T(N-2) + N + 1
 * = N * (N-1) * [N-2 * T(N-3) + 1] + N + 1
 * = N * (N-1) * (N-2) * T(N-3) + N *(N-1) + N + 1
 * = equivalent to N!
 * 
 * TC: O(2^N)
 * SC: O(1) - as we are using swapping
 */
public class hw_q2_All_Unique_Permutations {

    public ArrayList<ArrayList<Integer>> permute(ArrayList<Integer> A) {

        ArrayList<ArrayList<Integer>> list = new ArrayList<ArrayList<Integer>>();
        backtrack(0, A.size(), A, list);
        return list;

    }

    public void backtrack(int index, int n, ArrayList<Integer> A, ArrayList<ArrayList<Integer>> list) {

        // base condition
        if (index == n) {
            // only add unique combinations
            if (!list.contains(A)) {
                list.add(new ArrayList<Integer>(A));
            }
        }

        for (int i = index; i < n; i++) {
            // do: swap
            swap(A, index, i);
            backtrack(index + 1, n, A, list);
            // undo: swap
            swap(A, index, i);
        }
    }

    public void swap(ArrayList<Integer> A, int i, int k) {
        Integer temp = A.get(i);
        A.set(i, A.get(k));
        A.set(k, temp);
    }

    public static void main(String[] args) {

        hw_q2_All_Unique_Permutations t1 = new hw_q2_All_Unique_Permutations();

        // test case 1
        Integer[] A = { 1, 1, 2 };
        ArrayList<Integer> inputList = new ArrayList<Integer>(Arrays.asList(A));
        ArrayList<ArrayList<Integer>> outputList = t1.permute(inputList);
        System.out.println(outputList); // [[1, 1, 2], [1, 2, 1], [2, 1, 1]]

        // test case 2
        A = new Integer[] { 1, 2 };
        inputList = new ArrayList<Integer>(Arrays.asList(A));
        outputList = t1.permute(inputList);
        System.out.println(outputList); // [[1, 2], [2, 1]]

        // test case 3
        A = new Integer[] { 5, 6, 6, 5 };
        inputList = new ArrayList<Integer>(Arrays.asList(A));
        outputList = t1.permute(inputList);
        System.out.println(outputList); // [[1, 2], [2, 1]]
    }
}

File: my_problems/day72-dsa-backtracking-II/hw_q2_All_Unique_Permutations.js
/* All Unique Permutations */

//param A : array of integers
//return a array of array of integers
const permute = function (A) {

    let result = [];
    backtrack(A, 0, result);
    result = Array.from(new Set(result.map(JSON.stringify)), JSON.parse);
    console.log(result.length);
    return result;

    function backtrack(A, index, result) {
        if (index === A.length) {
            result.push([...A]);
        }

        for (let i = index; i < A.length; i++) {
            swap(A, i, index);
            backtrack(A, index + 1, result);
            swap(A, i, index);
        }
    }

    function swap(A, i, j) {
        let temp = A[i];
        A[i] = A[j];
        A[j] = temp;
    }
}

console.log(permute([1, 1, 2]));
console.log(permute([1, 2]));
console.log(permute([2, 5, 3, 4, 2, 5, 3, 1, 6]));